retrieval/discovery/ecs.go

```diff
@@ -0,0 +1,194 @@
+
+	// Get an initial set right away.
+	tg, err := ed.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ed.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ed *ECSDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	ecsService := ecs.New(session.New(), ed.aws)
+	ec2s := ec2.New(session.New(), ed.aws)
+
+	listServicesResponse, err := ecsService.ListServices(&ecs.ListServicesInput{
+		Cluster:    aws.String(ed.clusterName),
+		MaxResults: aws.Int64(100),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing ecs services: %s", err)
+	}
+
+	services, err := ecsService.DescribeServices(&ecs.DescribeServicesInput{
+		Cluster:  aws.String("default"),
+		Services: listServicesResponse.ServiceArns,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing describing services: %s", err)
+	}
+
+	for _, service := range services.Services {
+		// Possibly cache theses, since ARNS are immutable, afaik
+		taskDefinitionResponse, err := ecsService.DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
+			TaskDefinition: service.TaskDefinition,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error listing describing task description: %s", err)
+		}
+
+		monitored := false
+		monitoringPort := 0
+
+		var dockerLabels map[string]*string
+
+		// See if any container is exposing the specified ports
+		for _, containerDesc := range taskDefinitionResponse.TaskDefinition.ContainerDefinitions {
+			for _, portMapping := range containerDesc.PortMappings {
+				if *portMapping.HostPort >= ed.portFrom && *portMapping.HostPort <= ed.portTo {
+					monitored = true
+					monitoringPort = int(*portMapping.HostPort)
+					dockerLabels = containerDesc.DockerLabels
+					break
+				}
+			}
+		}
+
+		if !monitored {
+			continue
+		}
+
+		tasksArns, err := ecsService.ListTasks(&ecs.ListTasksInput{
+			ServiceName: aws.String(*service.ServiceName),
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error listing listing tasks: %s", err)
+		}
+
+		taskDescriptionsResponse, err := ecsService.DescribeTasks(&ecs.DescribeTasksInput{
+			Tasks: tasksArns.TaskArns,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error describing tasks: %s", err)
+		}
+
+		for _, task := range taskDescriptionsResponse.Tasks {
+			// Describe the container instance to get the IP
+			containerInstanceResponse, err := ecsService.DescribeContainerInstances(&ecs.DescribeContainerInstancesInput{
+				ContainerInstances: []*string{task.ContainerInstanceArn},
+			})
+			if err != nil {
+				return nil, fmt.Errorf("error describing container instance: %s", err)
+			}
+
+			containerInstance := containerInstanceResponse.ContainerInstances[0]
+			ec2Response, err := ec2s.DescribeInstances(&ec2.DescribeInstancesInput{
```

Comment:
been minimized so we do not do any redundant rpc calls. everything is batched and cached now.